torchtune/modules/transforms/tokenizers/_hf_tokenizer.py

```diff
@@ -146,3 +150,161 @@ def decode(self, token_ids: List[int]) -> str:
+            if content := token_info.get("content"):
+                special_tokens.add(content)
+
+    return list(special_tokens)
+
+
+class HuggingFaceModelTokenizer(ModelTokenizer):
+    def __init__(
+        self,
+        tokenizer_json_path: str,
+        *,
+        tokenizer_config_json_path: Optional[str] = None,
+        generation_config_path: Optional[str] = None,
+        truncation_type: str = "right",
+    ):
+        self.base_tokenizer = HuggingFaceBaseTokenizer(
+            tokenizer_json_path=tokenizer_json_path,
+            tokenizer_config_json_path=tokenizer_config_json_path,
+            generation_config_path=generation_config_path,
+        )
+
+        # Contents of the tokenizer_config.json
+        config = self.base_tokenizer.config
+
+        self.special_tokens = _infer_special_tokens_from_hf_config(config)
+        self.top_level_variables = self.extract_top_level_variables(config)
+
+        _env = jinja2.Environment(undefined=StrictUndefined)
+
+        # It is used sometimes in HF chat_templates
+        _env.globals["raise_exception"] = self._raise_helper
+
+        self.template = _env.from_string(
+            self._get_token_from_config(config, "chat_template")
+        )
+        self.truncation_type = truncation_type
+
+    def _raise_helper(self, msg):
+        raise Exception(msg)
+
+    def _get_token_from_config(self, config: Dict[str, Any], key: str) -> str:
+        """
+        HF BOS/EOS tokens are either stored as e.g. {'bos_token': 5}
+        or {'bos_token': {'content': 5, ...}}. This utility handles both.
+        """
+        token = config.get(key)
+        if isinstance(token, Dict):
+            if "content" not in token:
+                raise ValueError(f"Could not parse {key} from config")
+            token = token["content"]
+        else:
+            if not isinstance(token, str):
+                raise ValueError(f"Could not parse {key} from config")
+        return token
+
+    def extract_top_level_variables(self, config):
+        top_level = {}
+        for key, value in config.items():
+            if not isinstance(value, (dict, list)):
+                top_level[key] = value
+        return top_level
+
+    def tokenize_messages(
+        self,
+        messages: List[Message],
+        add_eos: bool = True,
+        max_seq_len: int | None = None,
+    ) -> Tuple[List[int], List[bool]]:
+        # This part is extremely hacky, but we need to handle case where we have variable access with jinja
+        special_tokens_mapping = {}
+        for token in self.special_tokens:
+            special_tokens_mapping[token] = self.base_tokenizer.encode(token)
+
+        tokenized_messages = []
+        mask = []
+        previous_tokens = []
+
+        for i, message in enumerate(messages):
+            current_messages = [
+                {"role": m.role, "content": m.content[0]["content"]}
+                for m in messages[: i + 1]
+            ]
```

Comment:
This seems a bit strange to me.. in my mind we should either be able to (a) render/tokenize all messages in one shot, or (b) loop over messages one at a time, render, tokenize, and concat. Why do we need to do this "cumulative tokenization"? (Is it because of the difficulties you mentioned with masking? If so I wonder whether there is an alternative)